src/foam/blob/RestBlobService.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package foam.blob;
+
+import foam.core.X;
+
+import java.net.URL;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.nio.ByteBuffer;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.InputStreamReader;
+
+public class RestBlobService
+    extends AbstractBlobService
+{
+  public static final int BUFFER_SIZE = 8192;
+
+  protected String address_;
+
+  public RestBlobService(String address) {
+    this.address_ = address + "/httpBlobService";
+  }
+
+  public RestBlobService(foam.core.X x, String address) {
+    setX(x);
+    this.address_ = address + "/httpBlobService";
+  }
+
+  public String getAddress() {
+    return address_;
+  }
+
+  @Override
+  public Blob put_(X x, Blob blob) {
+    if ( blob instanceof IdentifiedBlob ) {
+      return blob;
+    }
+    HttpURLConnection connection = null;
+    OutputStream os = null;
+    InputStream is = null;
+    IdentifiedBlob result = null;
+    try {
+      URL url = new URL(address_);
+      connection = (HttpURLConnection) url.openConnection();
+      
+      //configure HttpURLConnection
+      connection.setConnectTimeout(5 * 1000);
+      connection.setReadTimeout(5 * 1000);
+      connection.setDoOutput(true);
+      connection.setUseCaches(false);
+
+      //set request method
+      connection.setRequestMethod("PUT");
+
+      //configure http header
+      connection.setRequestProperty("Accept", "*/*");
+      connection.setRequestProperty("Connection", "keep-alive");
+      connection.setRequestProperty("Content-Type", "application/octet-stream");
+
+      //output blob into connection
+      long chunk = 0;
+      long size = blob.getSize();
```

Comment:
This is fine if the size of the blob is known but there are some cases where it is not. For example if you were to use an InputStreamBlob then this would not work.

Reply:
You mean there are some blob that we do not know the size. When we call blob.getSize(), it returns nothing? I check InputStreamBlob. We can get the size of a InputStreamBlob.

Reply:
Yes, but the InputStreamBlob implementation needs to be reworked because it crashes when you pass in an InputStream that contains too much data that can be stored in a byte array.

Reply:
If we do not know the size of a blob, how can we know how many iterations we need in order to get all bytes from a blob. Even though we can pass an empty buffer to the blob.read(), then check the return length of the buffer and put it in the while loop condition. It throws a IndexOutOfBoundsException, when we try to read the last chunk of the data.